### PR TITLE
HDDS-2579. Ozone client should refresh pipeline info if reads from al…

### PR DESCRIPTION
**Pull request by avijayanhwx**
…l Datanodes fail.

## What changes were proposed in this pull request?

Currently, if the client reads from all Datanodes in the pipleine fail, the read fails altogether. There may be a case when the container is moved to a new pipeline by the time client reads. In this case, the client should request for a refresh pipeline from OM, and read it again if the new pipeline returned from OM is different.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2579

## How was this patch tested?
Added unit tests.
Manual testing pending.

### CODE REVIEW

**Comment by xiaoyuyao** [comment]
Taking a look at the failed acceptance test. Looks like the latest docker image is the problem (last update 6 hr ago). cc: @elek , can you help taking a look? 

```
Status: Downloaded newer image for flokkr/hadoop:3.1.2

Output:  /tmp/smoketest/hadoop31/result/robot-hadoop31-hadoop31-createmrenv-scm.xml

sudo: apk: command not found
168
ERROR: Test execution of /home/runner/work/hadoop-ozone/hadoop-ozone/hadoop-ozone/dist/target/ozone-0.5.0-SNAPSHOT/compose/ozone-mr/hadoop31 is FAILED!!!!
```

**Comment by adoroszlai** [comment]
> Taking a look at the failed acceptance test. Looks like the latest docker image is the problem (last update 6 hr ago).
> 
> ```
> Status: Downloaded newer image for flokkr/hadoop:3.1.2
> ...
> sudo: apk: command not found
> ```

Thanks @xiaoyuyao for spotting this.  Filed [HDDS-2774](https://issues.apache.org/jira/browse/HDDS-2774).

**Comment by avijayanhwx** [comment]
@mukul1987 , @hanishakoneru Please review.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 1]
```diff
@@ -1031,7 +1032,33 @@ public OzoneInputStream readFile(String volumeName, String bucketName,
...
+        OmKeyArgs omKeyArgs = new OmKeyArgs.Builder()
+            .setVolumeName(omKeyInfo.getVolumeName())
+            .setBucketName(omKeyInfo.getBucketName())
+            .setKeyName(omKeyInfo.getKeyName())
+            .setRefreshPipeline(true)
+            .setSortDatanodesInPipeline(topologyAwareReadEnabled)
+            .build();
+        return ozoneManagerClient.lookupKey(omKeyArgs);
+      } catch (IOException e) {
+        LOG.error("Unable to get key!");
```
[the remark this answers is not in the document]
Comment (avijayanhwx) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java]: Sure, will do. Thanks for the review @hanishakoneru

**Comment by hanishakoneru** [comment]
Thanks @avijayanhwx.
+1. I will merge the PR.

**Comment by avijayanhwx** [comment]
Thanks for the review and commit @hanishakoneru.

**Pull request merged**